src/libraries/System.Linq/src/System/Linq/Iterator.SizeOpt.cs

```diff
@@ -0,0 +1,97 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+
+namespace System.Linq;
+
+public static partial class Enumerable
+{
+    /// <summary>
+    /// An iterator that implements <see cref="IList{T}"/>. This is used primarily in size-optimized
+    /// code to turn linear-time iterators into constant-time iterators.  The primary cost is
+    /// additional type checks, which are small compared to generic virtual calls.
+    /// </summary>
+    private sealed class SizeOptIListSelectIterator<TSource, TResult>(IList<TSource> _source, Func<TSource, TResult> _selector)
+        : Iterator<TResult>, IList<TResult>
```

Comment:
We have purposefully not implemented IList on the corresponding select iterators in LINQ because it causes behavioral changes that were deemed too problematic (we actually tried this a few releases ago and had to revert). If memory serves, the issue was that folks were relying on methods like Count to invoke side-effecting delegates in Selects, e.g.
```C#
Enumerable.Range(1, 10).Select(i =>
{
    Console.WriteLine(i);
    return i;
}).Count();
```
and if the Select iterator is itself an IList, then checks in methods like Count will short-circuit the operation and fail to perform those side effects.